### ISSUE TITLE
Tab index on tables

### ISSUE THREAD

**Issue by toni-sharpe** [Sage/carbon#815]
Currently if data is stored in a table and the rows are links to further data the user cannot tab to these elements and trigger the click action using the keyboard

It's probable that `tabindex` should be added to `TableRow` or `TableCell` (or both in fact) defaulting to off on the component but defaulting to on for rows in a basic `TableAjax`

UX input would be appreciated here, concerning what defaults should be and how cells and rows should be handled @benwilsonsage

**Comment by ianoxley** [comment]
Screen readers should give users access to the rows / columns without the need for a `tabindex` attribute on each row.

e.g. In VoiceOver bring up the rotor using ctrl+option+u.
Use left / right arrows to find Tables, then use the up / down arrows to select a table
Once the table is selected you can navigate the rows / columns using ctrl+option+arrow keys

You should be able to reach any naturally focusable elements such as links, form elements, buttons, etc. within the table via the keyboard already.

**Comment by andrewjtait** [comment]
This sounds like it needs to be resolved on a case by case basis, rather than a fix in the Table component? Should this be closed?

**Comment by toni-leigh** [comment]
Specifically here there is a problem with focus, when I tab through the UI and get to a table, i focus on the checkboxes at the edge but not on the rows, so I can't access them with the keyboard, I have to click on the mouse - I think this is a problem that should be fixed in Carbon, even if it's just putting a basic tabbability on the same element as the `onClick` for loading the table rows.

Personally I think the tables (grids) need to be more keyboard responsive, arrows for example would be great, as well as tab focus.

**Comment by ianoxley** [comment]
I don't think there's anything to fix in the `Table` component.

What @toni-leigh is describing sounds a lot like a [grid](https://www.w3.org/TR/wai-aria-practices/#grid), which would probably mean a new `Grid` component.

**Comment by toni-leigh** [comment]
@ianoxley et al. The gif should help - see vimium highlight the links, then the focus move along the headings then the checkboxes - neither hit the table rows, which are also clickable, as these things are not registered as such in the HTML

![focus](https://user-images.githubusercontent.com/10499070/29616535-be29b5b0-8809-11e7-8fd7-59f6cfc5938e.gif)

**Comment by andrewjtait** [comment]
so fix would be to highlight rows when they are clickable/links

**Comment by toni-leigh** [comment]
To provide `tabIndex` focusability on the same element as the `onClick` would fix it, I think, the gif shows the onclickability highlighted by hover and the vimium shows that this is not associated with keyboard accessibility

**Comment by ianoxley** [comment]
A better fix would be to add a link to the each row. That way links would show up in screen readers too:

<img width="772" alt="links" src="https://user-images.githubusercontent.com/52329/30052599-e0fe4db2-921d-11e7-921f-c24a0e933488.png">

**Comment by toni-leigh** [comment]
Absolutely @ianoxley anything that gets it out of mouse only navigation

**Comment by andrewjtait** [comment]
Link would be better for accessibility, but perhaps we need a combination of both to also cater for keyboard users? (I don't think UX would like a visible link in the row)

**Comment by toni-leigh** [comment]
You will need a visible focus state, but that doesn't have to be text link styling

**Comment by ianoxley** [comment]
The other thing to bear in mind is that if you've got a table with hundreds of rows then that could be a lot of tabbing for keyboard users.

**Comment by ianforster-sage** [comment]
Keyboard navigation for Table is now compliant with the DLS. Closing.
